adding docker buildx functionality into CI

xUnholy opened the pull request:
Signed-off-by: Michael Fornaro <20387402+xUnholy@users.noreply.github.com>

Completing integration of `docker buildx` work into the Github workflow

@sozercan if you don't mind taking a look and reviewing again 👍

xUnholy commented:
@sozercan not sure why this step has failed, it doesn't appear to be related to these changes but might need a second pair of eyes to confirm.

sozercan commented on .github/workflows/workflow.yaml:
```diff
@@ -104,8 +104,7 @@ jobs:
             make docker-login
-            make docker-build IMG=gatekeeper-e2e:latest
-            make docker-push-dev IMG=gatekeeper-e2e:latest DEV_TAG=${GITHUB_SHA::7}
+            docker-buildx-dev IMG=gatekeeper-e2e:latest DEV_TAG=${GITHUB_SHA::7}
```
`test` is part of `docker-build` too

sozercan commented on .github/workflows/workflow.yaml:
```diff
@@ -104,8 +104,7 @@ jobs:
             make docker-login
-            make docker-build IMG=gatekeeper-e2e:latest
-            make docker-push-dev IMG=gatekeeper-e2e:latest DEV_TAG=${GITHUB_SHA::7}
+            docker-buildx-dev IMG=gatekeeper-e2e:latest DEV_TAG=${GITHUB_SHA::7}
```
passing in `IMG` doesn't apply anymore

xUnholy replied:
Just confirming, `DEV_TAG` is still required though?

xUnholy replied:
This would also be true for `VERSION` in the release

sozercan commented:
@xUnholy I believe CI was a fluke (hopefully fixed by #601)

sozercan commented:
since image building takes much longer with this change, we should check if we need to increase timeouts in CI for pre-release and release

sozercan commented on Makefile:
```diff
@@ -181,7 +181,7 @@ docker-build: test
 # Experimental docker feature to build cross platform multi-architecture docker images
 # https://docs.docker.com/buildx/working-with-buildx/
-docker-buildx-dev:
+docker-buildx-dev: test
```
add `test` for `docker-buildx-release` also

xUnholy replied:
Addressed 👍

xUnholy commented:
@sozercan everything appears to look in order

sozercan commented on .github/workflows/workflow.yaml:
```diff
@@ -104,8 +104,7 @@ jobs:
             make docker-login
-            make docker-build IMG=gatekeeper-e2e:latest
-            make docker-push-dev IMG=gatekeeper-e2e:latest DEV_TAG=${GITHUB_SHA::7}
+            docker-buildx-dev DEV_TAG=${GITHUB_SHA::7}
```
```suggestion
            make docker-buildx-dev DEV_TAG=${GITHUB_SHA::7}
```

xUnholy replied:
Good find, I'll fix that.

sozercan commented:
Sorry about the delay, wanted to test this to make sure we are still within the timeouts.

There a few additional changes:

- Dockerfile needs to have `--platform=$BUILDPLATFORM` otherwise it doesn't seem to find the correct base image for the platform. So it'll look like:

```
FROM --platform=$BUILDPLATFORM golang:1.13-alpine as builder
```

- Since we added `--platform` we need to change the existing `docker-build` in the Makefile to `docker-buildx`, and it needs `--load` instead of `--push` since we are just running e2e tests on this, and only produce `linux/amd64` since docker cannot save multi-arch manifests locally. Also, update `e2e-build-load-image` to call `docker-buildx` instead.
That'll look something like:

```
docker-buildx: test
	@if ! docker buildx ls | grep -q container-builder; then\
		docker buildx create --platform "linux/amd64,linux/arm64,linux/arm/v7" --name container-builder --use;\
	fi
	docker buildx build --platform "linux/amd64" \
		-t $(IMG) \
		. --load
```

and 

```
e2e-build-load-image: docker-buildx
	kind load docker-image --name kind ${IMG}
```

xUnholy commented:
> Dockerfile needs to have --platform=$BUILDPLATFORM otherwise it doesn't seem to find the correct base image for the platform.

This isn't exactly correct, this isn't required because we're compiling with GO environment variables to set the GOOS, GOARCH, GOARM and this is just in the builder stage of the image which isn't used in the end. I've been running this image locally for a while with no problems on my `arm64` device.

I updated the other feedback, I also added a CI step which is a cleaner way to enable docker buildx experiment features, so probably don't need the `DOCKER_CLI_EXPERIMENTAL=enabled` although I'll leave it to make local development easy.

sozercan commented:
It does work for me locally, but getting `runtime execution on platform linux/arm/v7 not supported`  error without `--platform=$BUILDPLATFORM` in GH Actions: https://github.com/sozercan/gatekeeper/runs/686219733?check_suite_focus=true
Here it is building with `--platform=$BUILDPLATFORM`: https://github.com/sozercan/gatekeeper/runs/686236965?check_suite_focus=true

sozercan commented on .github/workflows/workflow.yaml:
```diff
@@ -104,30 +104,35 @@ jobs:
       - name: Check out code into the Go module directory
         uses: actions/checkout@v2
 
+      - name: Set up Docker Buildx
```
why is this only in tagged-release? doesn't make `docker-buildx-release` already handle this for us since it does create the builder and it'll get overridden since it has `--use`

xUnholy replied:
The Makefile ref is so for reproducibility when running locally. We do need to add this to the other job though.

sozercan commented:
Looks like because it is using qemu, it is adding about 10-12 mins to build time. is it possible to avoid this?

cpuguy83 commented on Makefile:
```diff
@@ -177,24 +180,26 @@ docker-push-release:  docker-tag-release
-	docker build --pull . -t ${IMG}
+docker-buildx: test
+	@if ! docker buildx ls | grep -q container-builder; then\
+		docker buildx create --platform "linux/amd64,linux/arm64,linux/arm/v7" --name container-builder --use;\
```
You should not need to specify platforms here.
When doing cross compilation the build platform is linux/amd64 no matter what the target platform is. This is specifying the build platform.

cpuguy83 replied:
Inside your Dockerfile you need to set `FROM --platform=$BUILDPLATFORM`, unless you want to do native compilation (instead of cross compile).

xUnholy replied:
@cpuguy83 if you look at the Dockerfile there is no need for adding `--platform=$BUILDPLATFORM` as mentioned in my previous reply, the final base is a distroless and we're simply coping the go binary in its compiled form to the final stage, there is no point adding that to the builder stage as golang doesn't require it in this context.

cpuguy83 replied:
buildkit will try to build using the platform specified as the target arch.
`FROM --platform=$BUILDPLATFORM` tells it to use the image for the build platform (That is the arch/os of the build host) instead of the target platform.

cpuguy83 replied:
This would go on your `builder` image, not the final.

xUnholy replied:
There is no value in putting it into the builder stage because go doesn't require the host OS to be multi-arch as long as the inputs are validated and correct which are passed from `TARGETPLATFORM` then it is sufficient as illustrated by this large community who maintain multi-arch images specifically for armv7l and arm64 https://github.com/raspbernetes/multi-arch-images .

Hope that makes sense? We don't need the builder image to be multiarch.

cpuguy83 replied:
The reason to do this is so you don't end up building in qemu, and instead build natively on the host arch with cross compilation.
Again, without setting the platform for the build image you will get qemu (or whatever is configured for binfmt), not cross compilation.

xUnholy commented:
> Looks like because it is using qemu, it is adding about 10-12 mins to build time. is it possible to avoid this?

The only idea I have for potentially speeding this up would be to put the building into a different job that runs in parallel and just adding a `needs:` clause for any required other jobs, then changing `--push` to `--load` and only pushing on successful test jobs.

Did you have any thoughts on speeding up the build time? Are there any other potential small wins that could be made within the workflow?

sozercan commented:
Can we try with `--platform=$BUILDPLATFORM` (without qemu) to check if that speeds up build time and still produces working ARM builds?

xUnholy commented:
@sozercan @cpuguy83 I'll add the fields and we can check what kind of performance improvements we gain 👍

sozercan commented:
@xUnholy apologies, missed that this was updated.

container build times looks good to me now:
https://github.com/sozercan/gatekeeper/runs/732977547?check_suite_focus=true

we can remove the `ghaction-docker-buildx` now, since we are building multi arch in Go and don't need qemu and buildx instance is created by `docker-buildx` in Makefile.

xUnholy commented:
@sozercan updated PR to reflect requested changes.
